File: 06ros/ws02/second_ws/src/use_cus2_srv_pkg/src/server_cus2_srv.cpp
//
// Created by zcb on 5/4/21.
//

#include <ros/ros.h>
#include <iostream>
#include <custom2_srvs/GenStudentId.h>


bool callback(custom2_srvs::GenStudentIdRequest &request, custom2_srvs::GenStudentIdResponse &response) {
    response.id = "000" + request.stu.name + "111" + std::to_string(request.stu.age) + "000";

    return true;
}


int main(int argc, char **argv) {

    std::string nodeName = "server_node";
    ros::init(argc, argv, nodeName);

    ros::NodeHandle node;

    std::string serviceName = "/zcb01/service";
    ros::ServiceServer server = node.advertiseService(serviceName, callback);

    ros::spin();


    return 0;
}

File: 06ros/ws02/second_ws/src/use_cus2_srv_pkg/scripts/server_cus2_srv.py
#! /usr/bin/env python
# coding=utf8

import rospy
from custom2_srvs.srv import GenStudentId, GenStudentIdRequest, GenStudentIdResponse


def callback(request):
    if not isinstance(request, GenStudentIdRequest):
        return
    response = GenStudentIdResponse()

    response.id = "000 %s 111 %d 000" % (request.stu.name, request.stu.age)

    return response


if __name__ == '__main__':
    nodeName = "server_node"
    rospy.init_node(nodeName)

    serviceName = "/zcb01/service"
    server = rospy.Service(serviceName, GenStudentId, callback)  # server

    rospy.spin()
